library/globnames.ml

```diff
@@ -43,13 +48,13 @@ let subst_global_reference subst ref = match ref with
     let ind' = subst_ind subst ind in
       if ind==ind' then ref else IndRef ind'
   | ConstructRef ((kn,i),j as c) ->
-    let c',t = subst_constructor subst c in
+    let c' = subst_constructor_reference subst c in
       if c'==c then ref else ConstructRef c'
 
 let subst_global subst ref = match ref with
   | VarRef var -> ref, mkVar var
   | ConstRef kn ->
-     let kn',t = subst_con_kn subst kn in
+     let kn',t = subst_con subst (Univ.in_punivs kn) in
```

Comment:
This should have some `XXX` comment.

Reply:
It is unclear what is the intended effect of the substitution wrt universes. If I have a module depending on a polymorphic name and the name is substituted by a name (which is the common case), do I want all the occurrences of the name to share the same universe instance or do I want each occurrence to have fresh universes?

Then, if I substitute a name by a term (due to `Inline`), do I want each occurrence to be substituted by an instance with new fresh universes? I would say yes. But if yes, this means that we should change the type of subst_global so that it returns not a `constr` but the closure of a `constr` over its universe parameters.

@mattam82, what do you think?

Reply:
I talked too quickly. Reading the discussions, #8974 seems a very good direction.